IBAN validation + test added

philsweb opened the pull request:
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

The best way to propose a feature is to open an issue first and discuss your ideas there before implementing them.

Always follow the [contribution guidelines](https://github.com/cakephp/cakephp/blob/master/.github/CONTRIBUTING.md) when submitting a pull request. In particular, make sure existing tests still pass, and add tests for all new behavior. When fixing a bug, you may want to add a test to verify the fix.

saeideng committed:
add here to the return of getAttributes

markstory committed:
Merge pull request #12089 from cakephp/req-here

add here to the return of getAttributes

Commit:
IBAN validation + test added

Commit:
IBAN validation comments resolved

stickler-ci committed:
Fixing style errors.

dereuromark commented on src/Validation/Validation.php:
```diff
@@ -181,25 +187,27 @@ public static function blank($check)
     public static function cc($check, $type = 'fast', $deep = false, $regex = null)
     {
-        if (!is_scalar($check)) {
+        if (! is_scalar($check)) {
```
Why are you changing this to an invalid CS? There must be no space after casts.
Also, please only change exactly the lines you want to change, no other noise changes.

philsweb replied:
Sure. Sorry, my bad

dereuromark commented on src/Validation/Validation.php:
```diff
@@ -82,8 +83,8 @@ class Validation
     protected static $_pattern = [
-        'hostname' => '(?:[_\p{L}0-9][-_\p{L}0-9]*\.)*(?:[\p{L}0-9][-\p{L}0-9]{0,62})\.(?:(?:[a-z]{2}\.)?[a-z]{2,})',
-        'latitude' => '[-+]?([1-8]?\d(\.\d+)?|90(\.0+)?)',
+        'hostname'  => '(?:[_\p{L}0-9][-_\p{L}0-9]*\.)*(?:[\p{L}0-9][-\p{L}0-9]{0,62})\.(?:(?:[a-z]{2}\.)?[a-z]{2,})',
```
Inline alignment is evil. We don't do that. :)

Commit:
IBAN validation + test, noise changes removed

Commit:
Merge branch 'master' of https://github.com/philsweb/cakephp

# Conflicts:
#	src/Validation/Validation.php
#	tests/TestCase/Validation/ValidationTest.php

Commit:
Discussions resolved

Double spaces removed

Merge branch 'master' of https://github.com/philsweb/cakephp

# Conflicts:
#	src/Validation/Validation.php
#	tests/TestCase/Validation/ValidationTest.php

Commit:
Double spaces removed

Commit:
Removing remaining double spaces

stickler-ci committed:
Fixing style errors.

dereuromark commented on src/Validation/Validation.php:
```diff
@@ -1559,6 +1559,41 @@ public static function hexColor($check)
+    {
+        $iban = str_replace(' ', '', $check);
+
+        if (preg_match('/^[A-Z]{2}[0-9]{2}[A-Z0-9]{1,30}$/i', $iban)) {
```
If you invert you can bail early here.

philsweb replied:
Sorry, not clear

dereuromark replied:
```php
if (!preg_match)) {
    return false;
}

... //continue without extra indentation level
```

See CakePHP coding guidelines.

philsweb replied:
Thanks!

dereuromark replied:
Sure you inverted properly?

dereuromark commented on src/Validation/Validation.php:
```diff
@@ -1552,19 +1620,58 @@ public static function isScalar($value)
+                $checksum %= 97;
+            }
+
+            return ((98 - $checksum) == $check);
```
Can we please use strict `===`?

philsweb replied:
Sure

Commit:
Strict comparison added

dereuromark commented:
Please squash all commits into a single one when you are done. There was a lot of noise changes that otherwise make 3.next/4.x etc merges more difficult and conflicts more likely where not necessary.

dereuromark commented:
Small side note for you:
You don't have your git set up correctly (email etc), so your name here is not clickable.

Commit:
Obsolete condition intending removed

Commit:
Inverting fixed

Commit:
Merge branch 'master' of https://github.com/philsweb/cakephp

# Conflicts:
#	src/Validation/Validation.php

philsweb commented:
@dereuromark Sorry, is it OK if  I open a new Pull Request right now? Cannot succeed in proper squashing.

dereuromark commented:
It is fine, we can sqash on merge afaik.

philsweb commented:
@dereuromark Thanks for your time, very first experience of contributing

inoas commented on src/Validation/Validation.php:
```diff
@@ -1559,6 +1559,40 @@ public static function hexColor($check)
     }
 
+    /**
+     * Check that the input value is a valid IBAN number
```
one number too much:

IBAN number = "International Bank Account Number number"

inoas commented on src/Validation/Validation.php:
```diff
@@ -1559,6 +1559,40 @@ public static function hexColor($check)
+     * @param string $check The value to check
+     * @return bool Success
+     */
+    public static function iban($check)
```
https://www.iban.com/structure.html

This validation method will pass IBANs that have wrong country codes or country code missmatches.
I am not saying that this should be fixed but the doc header should quote what the method does and what not.
I'd also say it should be `ibanSyntax($value)`.

The method checks for the country code to exist, the checksum to exist and then checks the body with Luhn algorithm

dereuromark replied:
iban() should suffice, the syntax sounds more like spacing check (or length only) or sth.

inoas replied:
it does nothing more than that. it does not validate an iban(). If it were it would be required to check the country code against the length, and possibly even the national bank codes against a database https://en.wikipedia.org/wiki/International_Bank_Account_Number#IBAN_formats_by_country

It is a pure stateless syntax check with. Validating an IBAN *for real* requires much more work. Thus `ibanSyntax()` hints you very clearly that you are only checking the form (uppercase, no whitespaces, max length, country code exists at right spot, checksum exists at right spot, body matches against checksum via Luhn)

dereuromark replied:
What about creditCard, ip ones, url etc?
They all also only do Syntax check, that is already very obvious here from the scope.

inoas commented on tests/TestCase/Validation/ValidationTest.php:
```diff
@@ -3172,4 +3172,19 @@ public function testHexColor()
+     */
+    public function testIban()
+    {
+        $this->assertTrue(Validation::iban('AD1200012030200359100100'));
```
Could this include all of these:
https://www.iban.com/structure.html

listed as IBAN SEPA and IBAN countries?

philsweb replied:
Done

inoas commented on src/Validation/Validation.php:
```diff
@@ -1559,6 +1559,41 @@ public static function hexColor($check)
+     */
+    public static function iban($check)
+    {
+        $iban = strtoupper(str_replace(' ', '', $check));
```
AFAIK IBANs are always uppercase and expected to be so. I'd not allow lowercase, instead strtoupper should happen on the data supplied by - say - the user.

philsweb commented:
@inoas Will deal with that in a while, thanks

Commit:
@inoas comments resolved

Commit:
Double spaces removed

inoas commented on src/Validation/Validation.php:
```diff
@@ -1559,6 +1559,41 @@ public static function hexColor($check)
+    /**
+     * Check that the input value has a valid International Bank Account Number IBAN syntax
+     * Requirements are uppercase, no whitespaces, max length 34, country code and checksum exist at right spots,
+     * body matches against checksum via Luhn algorithm
```
:+1: for single source of truth/error.

ADmad replied:
There's already a `Validation::luhn()` :slightly_smiling_face:

makallio85 replied:
I guess the algorithms are different

Commit:
IBAN validation description updated

markstory commented:
Thank you @philsweb :clap:
